lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1357,6 +1398,278 @@ class DifferentiableActivityInfo {
+
+           // Check if there are any active results or arguments. If not, skip
+          // this instruction.
+          auto hasActiveResults = llvm::any_of(
+              allResults, [&](SILValue res) {
+            return activityInfo.isActive(res, indices);
+          });
+          auto hasActiveArguments = llvm::any_of(
+              ai->getArgumentsWithoutIndirectResults(), [&](SILValue arg) {
+            return activityInfo.isActive(arg, indices);
+          });
+          if (!hasActiveResults || !hasActiveArguments)
+            continue;
+
+           unsigned source;
+          AutoDiffIndexSubset *parameters;
+
+           SmallVector<unsigned, 8> activeParamIndices;
+          SmallVector<unsigned, 8> activeResultIndices;
+          collectMinimalIndicesForFunctionCall(
+              ai, allResults, indices, activityInfo, activeParamIndices,
+              activeResultIndices);
+          source = activeResultIndices.front();
+
+           // If function is already marked differentiable, differentiate WRT
+          // all parameters.
+          auto originalFnSubstTy = ai->getSubstCalleeType();;
+          if (originalFnSubstTy->isDifferentiable()) {
+            parameters = originalFnSubstTy->getDifferentiationParameterIndices();
+          } else {
+            parameters = AutoDiffIndexSubset::get(
+                original->getASTContext(),
+                ai->getArgumentsWithoutIndirectResults().size(),
+                activeParamIndices);
+          }
+          SILAutoDiffIndices curIndices(activeResultIndices.front(),
+          AutoDiffIndexSubset::get(
+              builder.getASTContext(),
+              ai->getArgumentsWithoutIndirectResults().size(),
+              activeParamIndices));
+
+           // Check and diagnose non-differentiable original function type.
+          auto diagnoseNondifferentiableOriginalFunctionType =
+              [&](CanSILFunctionType origFnTy) {
+                // Check and diagnose non-differentiable arguments.
+                for (unsigned paramIndex : range(origFnTy->getNumParameters())) {
+                  if (curIndices.isWrtParameter(paramIndex) &&
+                          !origFnTy->getParameters()[paramIndex]
+                          .getSILStorageType()
+                          .isDifferentiable(builder.getModule())) {
+                    return true;
+                  }
+                }
+                // Check and diagnose non-differentiable results.
+                if (!origFnTy->getResults()[curIndices.source]
+                        .getSILStorageType()
+                        .isDifferentiable(builder.getModule())) {
+                  return true;
+                }
+                return false;
+              };
+          if (diagnoseNondifferentiableOriginalFunctionType(originalFnSubstTy))
+            continue;
+
+           auto JVPType = originalFnSubstTy->getAutoDiffAssociatedFunctionType(
```

Comment:
```suggestion
           auto assocFnType = originalFnSubstTy->getAutoDiffAssociatedFunctionType(
```
`JVPXXX` -> `assocFnXXX` and `differentialXXX` -> `linearMapXXX` below as well please.